course5/homework06/simple_string.cc

```diff
@@ -0,0 +1,232 @@
+#include "simple_string.h"
+#include <memory>
+#include <stdexcept>
+#include <cstring>
+
+
+namespace simple_string
+{
+
+String::String() = default;
+
+String::String(const char* initString)
+{
+    l = std::strlen(initString);
+    buffer = std::make_unique<char[]>(l + 1);
+    std::strcpy(buffer.get(), initString);
+}
+
+String::String(const String& src) // copy
+{
+    l = src.l;
+
+    if (src.buffer)
+    {
+        buffer = std::make_unique<char[]>(l + 1);
+        std::strcpy(buffer.get(), src.buffer.get());
+    }
+}
+
+String::String(String&& src) // move
+{
+    buffer = std::move(src.buffer);
+    l = src.l;
+
+    src.buffer = nullptr;
+    src.l = 0;
+}
+
+String& String::operator=(const String& src) // copy assignment
+{
+    l = src.l;
+    if (src.buffer)
```

Comment:
So this is the corner case I was looking for.

If you assign empty string to a non-empty one, and the empty string has no buffer, then the assigned-to will keep its old content.